### PR TITLE
vmi-mutator: Set default for VMI guest memory

### LINKED ISSUE #10521: VMs without guest memory settings cannot start
**What happened**:

I created a VM without an instancetype nor any other settings related to guest memory. The VM was created successfully (it passed vm-mutator/vm-admitter), but it cannot be started, because a VMI cannot be created from it (VMI does not pass vmi-admitter).

**What you expected to happen**:
I expected the created VM to be able to start. When no memory is specified, the vmi-mutator should set a default value (e.g. `512Mi` and similar to CPUs, e.g. if no CPU is provided we default to 1 socket, 1 core, 1 thread).

**How to reproduce it (as minimally and precisely as possible)**:
- Create VM with provided YAML
- Run `virtctl start vm-cirros`
- Run `kubectl get vm vm-cirros -o yaml` to observe failure in status conditions

**Additional context**:

**Environment**:
- KubeVirt version (use `virtctl version`): 1.0.0
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications:

```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: vm-cirros
spec:
  running: false
  template:
    spec:
      domain:
        devices: {}
      volumes:
      - containerDisk:
          image: registry:5000/kubevirt/cirros-container-disk-demo:devel
        name: containerdisk

```

- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by 0xFelix**
**What this PR does / why we need it**:

This PR adds the DefaultGuestMemory field to KubeVirtConfiguration.
This value will be used to set a default amount of memory for VMIs, which have no memory
configured.

Follow up to https://github.com/kubevirt/kubevirt/pull/10450 and https://github.com/kubevirt/kubevirt/pull/10463

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10521 

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
None
```

### CODE REVIEW

**Review by EdDev — COMMENTED** [review]
Defaults, especially in the KV config, caused a lot of issues in the network area.
It mainly increased complexity in many areas of the code, many times causing bugs.

IMO, the pattern of having defaults defined this way is problematic.
It would be preferred if the values would come in through other means.

BTW, do I understand correctly that this is done through the mutation hook? i.e. the spec is changed?
Are there other fields we do this?

File: pkg/virt-api/webhooks/defaults.go [hunk 1]
```diff
@@ -236,3 +237,18 @@ func setDefaultArchitecture(clusterConfig *virtconfig.ClusterConfig, spec *v1.Vi
 		spec.Architecture = clusterConfig.GetDefaultArchitecture()
 	}
 }
+
+func setDefaultGuestMemory(clusterConfig *virtconfig.ClusterConfig, spec *v1.VirtualMachineInstanceSpec) {
+	if spec.Domain.Resources.Requests.Memory().Value() == 0 &&
+		(spec.Domain.Memory == nil || spec.Domain.Memory.Guest == nil && spec.Domain.Memory.Hugepages == nil) {
+		defaultGuestMemory := clusterConfig.GetConfig().DefaultGuestMemory
+		if defaultGuestMemory == nil {
+			parsedMemory := resource.MustParse("512Mi")
```
Comment (lyarwood) [inline pkg/virt-api/webhooks/defaults.go:246]: nit - We should move this into the actual KV CR definition with the `+default=`  marker in the future.

**Review by lyarwood — COMMENTED** [review]
> Defaults, especially in the KV config, caused a lot of issues in the network area. It mainly increased complexity in many areas of the code, many times causing bugs.

Were these defaults mutated into to a resource once or always looked up from the KV CR?

> IMO, the pattern of having defaults defined this way is problematic. It would be preferred if the values would come in through other means.

Such as? 

> BTW, do I understand correctly that this is done through the mutation hook? i.e. the spec is changed? Are there other fields we do this?

Yes much of the `VirtualMachineInstanceSpec` has associated defaults that we apply during mutation:

https://github.com/kubevirt/kubevirt/blob/0bacc588bd5c195f6a048581a151ad8d9f83e900/pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go#L81-L85

File: pkg/virt-api/webhooks/defaults.go (hunk 1 above)
Comment (lyarwood) [reply pkg/virt-api/webhooks/defaults.go:246]: For context here's some `kube-openapi` tests showing how the `+default=` marker can be used:

https://github.com/kubernetes/kube-openapi/blob/a781456278336f662dbccf613a9a725480448bce/test/integration/testdata/defaults/default.go

**Comment by lyarwood** [comment]
Looks like you missed a functional test asserting the previous behaviour:

`Tests Suite: [sig-compute]Configurations VirtualMachineInstance definition [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with no memory requested [test_id:3113]should failed to the VMI creation `

**Comment by EdDev** [comment]
> It is just a proposal, I also thought about making the default value non-configurable. Which other means would you propose?

Fail if someone has not set one (i.e. making it obligatory).
Or have a higher level abstraction doing this (e.g. templates).

> > Defaults, especially in the KV config, caused a lot of issues in the network area. It mainly increased complexity in many areas of the code, many times causing bugs.
> 
> Were these defaults mutated into to a resource once or always looked up from the KV CR?

The code is confusing (me), it appears as if the default network is applied both on the mutation [1] and on the vm controller [2]:

1. https://github.com/kubevirt/kubevirt/blob/ed6a2ab320733830b784df4ddb18af4c83e037be/pkg/virt-api/webhooks/defaults.go#L99
2. https://github.com/kubevirt/kubevirt/blob/eb96837b62048c96b82549fc825604241970cbe9/pkg/virt-controller/watch/vm.go#L1126

> > IMO, the pattern of having defaults defined this way is problematic. It would be preferred if the values would come in through other means.
> 
> Such as?

A higher level component that creates the VM/VMI with the "default" not requiring it as a must.

> > BTW, do I understand correctly that this is done through the mutation hook? i.e. the spec is changed? Are there other fields we do this?
> 
> Yes much of the `VirtualMachineInstanceSpec` has associated defaults that we apply during mutation:
> 
> https://github.com/kubevirt/kubevirt/blob/0bacc588bd5c195f6a048581a151ad8d9f83e900/pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go#L81-L85

Yea, now I see that even the default network may be handled there (although I do not understand why it appears in the controller as well).

Isn't this a problem with a potential race?
I mean, isn't it possible that others will create custom webhooks that may get triggered before "our" webhook? I remember vaguely that the order cannot be controlled (but maybe things changed or my memory fails me here).

**Comment by 0xFelix** [comment]
@lyarwood Updated to change the expectations of the failing test.

@EdDev Right now you may create a VM without memory, but you cannot start it (VMI creation fails). I would either say fail fast, like you say, and already fail the creation of the VM. Or be more tolerant about it and fill in a default value for the memory. E.g. for the CPU we do it too, so I think this is reasonable.

**Comment by lyarwood** [comment]
> > > Defaults, especially in the KV config, caused a lot of issues in the network area. It mainly increased complexity in many areas of the code, many times causing bugs.
> > 
> > Were these defaults mutated into to a resource once or always looked up from the KV CR?
> 
> The code is confusing (me), it appears as if the default network is applied both on the mutation [1] and on the vm controller [2]:
> 
> 1. https://github.com/kubevirt/kubevirt/blob/ed6a2ab320733830b784df4ddb18af4c83e037be/pkg/virt-api/webhooks/defaults.go#L99
> 2. https://github.com/kubevirt/kubevirt/blob/eb96837b62048c96b82549fc825604241970cbe9/pkg/virt-controller/watch/vm.go#L1126

Yeah the VirtualMachine controller needs to set these defaults on VirtualMachineInstances it is submitting so we can also apply preferences ahead of submission, for example a preferred interface model for the default interface.

We also need to support users creating VirtualMachineInstances directly thus the slight duplication here.

> > > IMO, the pattern of having defaults defined this way is problematic. It would be preferred if the values would come in through other means.
> > 
> > Such as?
> 
> A higher level component that creates the VM/VMI with the "default" not requiring it as a must.

We could move all of the defaulting out of the mutation webhook into the VirtualMachine controller but you'd then end up in a situation where current users of VirtualMachineInstances would see different behaviour during submission. I don't think this is a change we could land while the API is v1 but maybe something for v2 if that was ever to happen.

> > > BTW, do I understand correctly that this is done through the mutation hook? i.e. the spec is changed? Are there other fields we do this?
> > 
> > Yes much of the `VirtualMachineInstanceSpec` has associated defaults that we apply during mutation:
> > https://github.com/kubevirt/kubevirt/blob/0bacc588bd5c195f6a048581a151ad8d9f83e900/pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go#L81-L85
> 
> Yea, now I see that even the default network may be handled there (although I do not understand why it appears in the controller as well).

As above.

> Isn't this a problem with a potential race? I mean, isn't it possible that others will create custom webhooks that may get triggered before "our" webhook? I remember vaguely that the order cannot be controlled (but maybe things changed or my memory fails me here).

Yeah I don't think the order of mutation webhooks can be controlled when several target the same resource, namespace etc but they are still documented as they best place to set defaults during submission.

**Comment by acardace** [comment]
@0xFelix CPU is a compressible resource (you just get throttled if you don't have enough), but memory is not. I wouldn't treat them equally.

I'd rather have KubeVirt fail fast rather than automatically assign some amount of memory.

Thanks for bringing this up by the way, I didn't realize you could create a VM without specifying memory.

**Comment by 0xFelix** [comment]
@acardace They are not equal, you are right. But from a `virtctl` UX point of view I would prefer to have a default for the memory instead of failing the creation. I would say at some point a user will notice if they need more memory.

**Comment by vladikr** [comment]
/hold

I don't understand why we are doing this.
From my point of view, explicitly setting memory is an absolute requirement for running a VM.

Memory is directly related to the boot image, which is also a requirement for running a VM.
Imagine that we will set a default memory which is not enough for running the operating system. The VM will be created, and scheduled but will not boot.
We will waste resources, scheduler, and controller cycles for no reason.

**Comment by 0xFelix** [comment]
The idea was to allow creation of a somewhat bootable VM, in case of ignoring an instancetype inference failure for a VM created by `virtctl create vm`. Prior to #10450, `virtctl create vm` defaulted to `512Mi` of memory, if no instancetype was specified. Since enabling inference by default (#10450) and introducing a failure policy for inference (#10463 ) this PR would be required to keep the previous behavior when inference fails.

In general I noticed that you can create VMs without specifying any memory (#10521), so I consider this a bug not directly related to virtctl.

**Comment by lyarwood** [comment]
@0xFelix did you want to switch this PR over to rejecting `VirtualMachines` without memory defined?

**Comment by 0xFelix** [comment]
@lyarwood Will open a new PR for that.

**Comment by 0xFelix** [comment]
Closed in favor of #10546

**Pull request closed without merging**